File: grodbDate/src/nu/mine/mosher/grodb/date/YMD.java
package nu.mine.mosher.grodb.date;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import nu.mine.mosher.time.Time;

/**
 * Represents a date, specified as a year, month, and day, allowing
 * for some values to be unknown. An unknown day or month is specified as zero.
 *
 * @author Chris Mosher
 */
public class YMD implements Serializable, Comparable<YMD>
{
	/*
	 * One-based year, month, and day.
	 * Gregorian calendar is assumed.
	 * Zero indicates that field is "unknown"
	 * Negative year means B.C.
	 */
	private int year;
	private int month;
	private int day;

	private transient int hash;
	private transient Time approx;



	/**
	 * @param year
	 */
	public YMD(final int year)
	{
		this(year,0,0);
	}

	/**
	 * @param year
	 * @param month
	 */
	public YMD(final int year, final int month)
	{
		this(year,month,0);
	}

	/**
	 * @param year
	 * @param month
	 * @param day
	 */
	public YMD(final int year, final int month, final int day)
	{
		this.year = year;
		this.month = month;
		this.day = day;

		init();
	}


	/**
	 * @param time
	 */
	public YMD(final Time time)
	{
		final GregorianCalendar cal = new GregorianCalendar();
    	cal.setGregorianChange(new Date(Long.MIN_VALUE));
    	cal.setTime(time.asDate());
    	this.year = cal.get(Calendar.YEAR);
    	this.month = cal.get(Calendar.MONTH)+1;
    	this.day = cal.get(Calendar.DAY_OF_MONTH);
    	init();
	}



	/**
	 * @return the day, or zero if unknown
	 */
	public int getDay()
    {
        return this.day;
    }

    /**
     * @return the month (1 means January), or zero if unknown
     */
    public int getMonth()
    {
        return this.month;
    }

    /**
     * @return the year
     */
    public int getYear()
    {
        return this.year;
    }

    public Time getExactTime()
    {
    	if (!isExact())
    	{
    		throw new IllegalStateException();
    	}

    	return this.approx;
    }

	// never display this to the user!
	public Time getApproxTime()
	{
		return this.approx;
	}

    public boolean isExact()
    {
    	return valid(this.year) && valid(this.month) && valid(this.day);
    }

	public static YMD getMinimum()
    {
    	return new YMD(-9999,1,1);
    }

    public static YMD getMaximum()
    {
    	return new YMD(9999,12,31);
    }



	@Override
	public boolean equals(final Object object)
	{
		if (!(object instanceof YMD))
		{
			return false;
		}

		final YMD that = (YMD)object;
		return
			this.year == that.year &&
			this.month == that.month &&
			this.day == that.day;
	}

    @Override
	public int hashCode()
    {
    	return this.hash;
    }

    @Override
    public String toString()
    {
    	final StringBuilder sb = new StringBuilder();
    	if (this.year != 0)
    	{
        	if (this.year < 0)
        	{
        		sb.append('-');
        	}
    		sb.append(String.format("%04d",Math.abs(this.year)));
    		if (this.month > 0)
    		{
        		sb.append(String.format("-%02d",this.month));
        		if (this.day > 0)
        		{
        			sb.append(String.format("-%02d",this.day));
        		}
    		}
    	}
    	return sb.toString();
    }

    public int compareTo(final YMD that)
    {
    	return this.approx.compareTo(that.approx);
    }

    private static boolean valid(final int i)
	{
		return i != 0;
	}

	private Time calcApprox()
	{
		int m = this.month;
		int d = this.day;

		// if month and day are missing, assume mid-year (July 3).
		if (m == 0 && d == 0)
		{
			m = 7;
			d = 3;
		}
		// if just day is missing, assume mid-month (the 15th).
		else if (d == 0)
		{
			d = 15;
		}

    	return createTime(this.year,m,d);
	}

	private static Time createTime(final int year, final int month, final int day)
	{
		final GregorianCalendar cal = new GregorianCalendar();
    	cal.setGregorianChange(new Date(Long.MIN_VALUE));

    	cal.set(year,month-1,day,12,0,0);
    	cal.set(Calendar.MILLISECOND,0);

    	return new Time(cal.getTime());
	}

    private int calcHash()
    {
    	return this.approx.hashCode();
    }

    private void writeObject(final ObjectOutputStream s) throws IOException
    {
    	s.defaultWriteObject();
    }

    private void readObject(final ObjectInputStream s) throws IOException, ClassNotFoundException
    {
    	s.defaultReadObject();

		init();
    }

	private void init()
	{
		this.approx = calcApprox();
		this.hash = calcHash();
	}
}

File: grojs/YMD.js
/**
 * @fileoverview
 * Defines the {@link YMD} (year/month/day) class.
 */

/**
 * @class
 * Represents a date, specified as a year, month, and day, allowing
 * for unknown month and/or day. An unknown day or month is specified as zero.
 * @requires Util
 * 
 * @constructor
 * @param {Number} y year -9999 to -1 (for BC), or 1 to 9999 (for AD)
 * @param {Number} m month 1-12 (0 or omitted means unknown)
 * @param {Number} d day 1-31 (0 or omitted means unknown)
 * @param {Boolean} circa
 * @param {Boolean} julian
 * @return new {@link YMD}
 * @type YMD
 */
function YMD(y,m,d,circa,julian) {
	Util.verifyType(this,"YMD");

	/**
	 * year (-9999 to -1, or 1 to 9999)
	 * @private
	 * @type Number
	 */
	this.year = parseInt(y,10);
	if (this.year <= -10000 || this.year == 0 || 10000 <= this.year) {
		throw new Error("invalid year: "+this.year);
	}

	/**
	 * month (1 = Jan) (0 = unknown)
	 * @private
	 * @type Number
	 */
	this.month = !m ? 0 : parseInt(m,10);
	if (this.month < 0 || 12 < this.month) {
		throw new Error("invalid month: "+this.month);
	}

	/**
	 * day of month (1-31) (0 = unknown)
	 * @private
	 * @type Number
	 */
	this.day = !d ? 0 : parseInt(d,10);
	if (this.day < 0 || 31 < this.day) {
		throw new Error("invalid day: "+this.day);
	}

	/**
	 * if this date is an approximation
	 * @private
	 * @type Boolean
	 */
	this.circa = !!circa;

	/**
	 * if this date SHOULD be DISPLAYED in Julian Calendar
	 * @private
	 * @type Boolean
	 */
	this.julian = !!julian;

	/**
	 * @private
	 * @type Date
	 */
	this.approx = this.calcApprox();
}

/**
 * Gets the day
 * @return the day, or zero if unknown
 * @type Number
 */
YMD.prototype.getDay = function() {
    return this.day;
};

/**
 * Gets the month
 * @return the month (1 means January), or zero if unknown
 * @type Number
 */
YMD.prototype.getMonth = function() {
    return this.month;
};

/**
 * Gets the year
 * @return the year, or zero if unknown. (negative means BC)
 * @type Number
 */
YMD.prototype.getYear = function() {
    return this.year;
};

/**
 * @return if this date should be show using the Julian calendar
 * (if so, the caller is responsible for converting it).
 * @type Boolean
 */
YMD.prototype.isJulian = function() {
	return this.julian;
};

/**
 * @return if this is an approximate date
 * @type Boolean
 */
YMD.prototype.isCirca = function() {
	return this.circa;
};

/**
 * Gets the exact <code>Date</code> represented by this {@link YMD},
 * assuming it is exact. Throws otherwise.
 * @return the <code>Date</code> representing this exact {@link YMD} (at noon, local time).
 * @type Date
 * @throws if this {@link YMD} is missing day or month
 */
YMD.prototype.getExactDate = function() {
	if (!this.isExact()) {
		throw new Error("missing month or day on date that was presumed to be exact");
	}

	return this.approx;
};

/**
 * Gets a <code>Date</code> that can be used as an approximation
 * of this {@link YMD} for computation purposes.
 * Never display this value to the user!
 * @return an approximate <code>Date</code> for this {@link YMD}
 * @type Date
 */
YMD.prototype.getApproxDate = function() {
	return this.approx;
};

/**
 * Gets if this {@link YMD} is exact.
 * @return <code>true</code> if exact
 * @type Boolean
 */
YMD.prototype.isExact = function() {
	return YMD.valid(this.month) && YMD.valid(this.day);
};

/**
 * Returns a new {@link YMD} representing January 1, 9999 BC.
 * @return Jan. 1, 9999 BC
 * @type YMD
 */
YMD.getMinimum = function() {
	return new YMD(-9999,1,1);
};

/**
 * Returns a new {@link YMD} representing December 31, AD 9999.
 * @return Dec. 31, AD 9999
 * @type YMD
 */
YMD.getMaximum = function() {
	return new YMD(9999,12,31);
};



/**
 * @return date in yyyy-mm-dd format
 * @type String
 */
YMD.prototype.toString = function() {
	var s;
	s = "";

	if (this.circa) {
		s += "c. ";
	}

	if (this.year == 9999 || this.year == -9999) {
		if (YMD.equal(this,YMD.getMaximum())) {
			return "(MAX)";
		}
		if (YMD.equal(this,YMD.getMinimum())) {
			return "(MIN)";
		}
	}
	s += Util.digint(this.year,4);
	if (this.month > 0) {
		s += "-"+Util.digint(this.month,2);
		if (this.day > 0) {
			s += "-"+Util.digint(this.day,2);
		}
	}
	return s;
};

/**
 * @private
 * @param {Number} i
 * @return if given number is not zero (not unknown)
 * @type Boolean
 */
YMD.valid = function(i) {
	return i != 0;
};

/**
 * Approximate date. Never display this to the user; only use
 * it for approximation in calculations (for example, for sorting).
 * @private
 * @return approximate date (useful if month or day are unknown)
 * @type Date
 */
YMD.prototype.calcApprox = function() {
	var dt;
	var m = this.month;
	var d = this.day;

	// if month and day are missing, assume mid-year (July 3).
	if (m == 0 && d == 0) {
		m = 7;
		d = 3;
	}
	// if just day is missing, assume mid-month (the 15th).
	else if (d == 0) {
		d = 15;
	}

	dt = new Date(0);
	dt.setUTCFullYear(this.year);
	dt.setUTCMonth(m-1);
	dt.setUTCDate(d);
	return dt;
};

/**
 * Checks two {@link YMD}s for equality.
 * @param {YMD} a not null/undefined
 * @param {YMD} b not null/undefined
 * @return if a and b are equal
 * @type Boolean
 */
YMD.equal = function(a,b) {
	return a.year==b.year && a.month==b.month && a.day==b.day;
};


/**
 * Compares two {@link YMD}s, for sorting.
 * @param {YMD} a not null/undefined
 * @param {YMD} b not null/undefined
 * @return -1:a&lt;b, 0:a=b, +1:a&gt;b
 * @type Number
 */
YMD.order = function(a,b) {
	return Util.dateOrder(a.getApproxDate(),b.getApproxDate());
};

/**
 * 
 * @param {Object} r
 * @return if parser result is a YMD type
 * @type Boolean
 */
YMD.isParsedYMD = function(r) {
	if (!r) {
		return false;
	}
	return r.hasOwnProperty("year") && r.hasOwnProperty("month") && r.hasOwnProperty("day");
};

/**
 * 
 * @param {Object} r result from parser
 * @return new {@link YMD} (or null if !r)
 * @type YMD
 */
YMD.fromParserResult = function(r) {
	if (!r) {
		return null;
	}
	return new YMD(r.year,r.month,r.day,r.approx,r.julian);
};
